src/gpu_util.cpp

```diff
@@ -264,6 +266,84 @@ void gpu_get_all_domain_id(std::vector<uint16_t>* pgpus_domain_id,
   }
 }
 
+/**
+ * Get PCI BDF (Bus, Device, Function) for all GPUs.
+ * @param pgpus_domain_id ptr to vector that will store PCI BDF
+ * @return void
+ */
+void gpu_get_all_pci_bdf(std::vector<std::string>* ppci_bdf) {
+
+  ifstream f_id, f_prop;
+  char path[KFD_PATH_MAX_LENGTH];
+  std::string prop_name;
+  int gpu_id;
+  uint32_t domain_val;
+  uint32_t location_val;
+  std::string domain;
+  std::string bus;
+  std::string device;
+
+  // Discover the number of nodes: Inside nodes folder there are only folders
+  // that represent the node number
+  int num_nodes = gpu_num_subdirs(KFD_SYS_PATH_NODES, "");
+
+  // get all GPUs device id
+  for (int node_id = 0; node_id < num_nodes; node_id++) {
+
+    snprintf(path, KFD_PATH_MAX_LENGTH, "%s/%d/gpu_id", KFD_SYS_PATH_NODES,
+        node_id);
+    f_id.open(path);
+
+    snprintf(path, KFD_PATH_MAX_LENGTH, "%s/%d/properties",
+        KFD_SYS_PATH_NODES, node_id);
+    f_prop.open(path);
+
+    f_id >> gpu_id;
+
+    if (gpu_id != 0) {
+
+      while (f_prop >> prop_name) {
+
+        if (prop_name == "domain") {
+
+          // Get domain string
+          f_prop >> domain_val;
+
+          std::stringstream ss;
+          ss << std::hex << domain_val;
+          domain =  ss.str();
+
+          domain = std::string(4 - std::min(4, (int)domain.length()), '0') + domain;
+
+          continue;
+        }
+        else if(prop_name == "location_id") {
+
+          // Get bus and device string
+          f_prop >> location_val;
+
+          std::stringstream ss;
+          ss << std::hex << location_val;
+          std::string location (ss.str());
+
+          location = std::string(4 - std::min(4, (int)location.length()), '0') + location;
+
+          bus = location.substr(0,2);
+          device = location.substr(2,2);
+
+          continue;
+        }
+      }
+
+      /* Form PCI BDF */
+      (*ppci_bdf).push_back(domain + ":" + bus + ":" + device + ".0" );
+    }
+
+    f_id.close();
```

Comment:
destructor closes it, so do we need explicit close?

Reply:
Do we have destructor here ? As the methods all are static right ?
Also, closing stream inside loop looks clearer and followed as convention in other similar APIs as well.

Reply:
Whole RAII ensures cleanup happens before we take object out of scope. methods are static but f_id is local variable in this scope and when it encounters braces its destructor is called for sure, which closes the stream object.

Reply:
In that case, f_id & f_prop have to be defined inside the for loop. right ?

Reply:
As discussed, for now will keep as it is.